prometheus client library panics with `counter cannot decrease in value`

### What happened?

We have unit test(s) that gather metrics for verification using `legacyregistry.DefaultGatherer`, for example: 
https://github.com/kubernetes/kubernetes/blob/0ae6ef69b8cf49af8d6e87fca3d400716bba631b/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go#L1288-L1291

We have seen test to panic with the following error:
```
panic: counter cannot decrease in value
```

the panic seems to be originating from prometheus client library:
```
goroutine 6151 [running]:
k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus.(*counter).Add(0xc0003a41e0, 0xc067600000000000)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus/counter.go:109 +0x1a5
k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus.(*goCollector).Collect(0xc0003ac000, 0xc0001e4f50)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus/go_collector_go117.go:147 +0x6a9
k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus/registry.go:446 +0x134
created by k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).Gather
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus/registry.go:538 +0x1138
```

using the stack trace, we can see the offending code site:
https://github.com/kubernetes/kubernetes/blob/master/vendor/github.com/prometheus/client_golang/prometheus/counter.go#L109
```
func (c *counter) Add(v float64) {
	if v < 0 {
		panic(errors.New("counter cannot decrease in value"))
	}
...
}
```
I checked the apf metrics package, we don't set any decreasing value for any counter, and it's clear from the stack trace that `Add` call is being issued by the collector.
https://github.com/kubernetes/kubernetes/blob/master/vendor/github.com/prometheus/client_golang/prometheus/go_collector_go117.go#L141-L149
```
		case *counter:
			// Guard against decreases. This should never happen, but a failure
			// to do so will result in a panic, which is a harsh consequence for
			// a metrics collection bug.
			v0, v1 := m.get(), unwrapScalarRMValue(sample.Value)
			if v1 > v0 {
				m.Add(unwrapScalarRMValue(sample.Value) - m.get())
			}
			m.Collect(ch)
```

so `m.Add(unwrapScalarRMValue(sample.Value) - m.get())` is producing a negative value here; looks like the vendored version of the collector is not thread safe, and the prometheus folks have already added lock for the 1.17 collector.
https://github.com/prometheus/client_golang/pull/975

https://github.com/prometheus/client_golang/pull/975/commits/648c419b2a7ba824f85d5c7526210d891373d55c#diff-6096ed186872e4be5a35c1c6e8fbee95be0f2268753ad7b1d9685242e5dc681bR141-R150

The reason I see the test `TestPriorityAndFairnessWithPanicRecoveryAndTimeoutFilter` fail with this panic is because its subtests (that gathers metric) run in parallel.

I think we should re-vendor the prometheus client library, any thoughts?


### What did you expect to happen?

No panic

### How can we reproduce it (as minimally and precisely as possible)?

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108013/pull-kubernetes-unit/1496512399061028864

--- a/go.mod
+++ b/go.mod
@@ -66,7 +66,7 @@ require (
 	github.com/opencontainers/selinux v1.8.2
 	github.com/pkg/errors v0.9.1
 	github.com/pmezard/go-difflib v1.0.0
-	github.com/prometheus/client_golang v1.12.0
+	github.com/prometheus/client_golang v1.12.1
 	github.com/prometheus/client_model v0.2.0
 	github.com/prometheus/common v0.32.1
 	github.com/quobyte/api v0.1.8
@@ -381,7 +381,7 @@ replace (
 	github.com/pmezard/go-difflib => github.com/pmezard/go-difflib v1.0.0
 	github.com/posener/complete => github.com/posener/complete v1.1.1
 	github.com/pquerna/cachecontrol => github.com/pquerna/cachecontrol v0.0.0-20171018203845-0dec1b30a021
-	github.com/prometheus/client_golang => github.com/prometheus/client_golang v1.12.0
+	github.com/prometheus/client_golang => github.com/prometheus/client_golang v1.12.1
 	github.com/prometheus/client_model => github.com/prometheus/client_model v0.2.0
 	github.com/prometheus/common => github.com/prometheus/common v0.32.1
 	github.com/prometheus/procfs => github.com/prometheus/procfs v0.7.3

--- a/staging/src/k8s.io/component-base/go.mod
+++ b/staging/src/k8s.io/component-base/go.mod
@@ -11,7 +11,7 @@ require (
 	github.com/google/go-cmp v0.5.5
 	github.com/matttproud/golang_protobuf_extensions v1.0.2-0.20181231171920-c182affec369 // indirect
 	github.com/moby/term v0.0.0-20210619224110-3f7ff695adc6
-	github.com/prometheus/client_golang v1.12.0
+	github.com/prometheus/client_golang v1.12.1
 	github.com/prometheus/client_model v0.2.0
 	github.com/prometheus/common v0.32.1
 	github.com/prometheus/procfs v0.7.3

